### PR TITLE
Fix: Replace internal docs action

### PR DESCRIPTION
**Pull request by andrew-causalens**
## Description and Motivation

Docs were previously built by the internal docs action. Because this repo is now open-source, it cannot use internal (closed source) actions - this is enforced by GitHub and cannot be worked around.

Instead the docs publishing step has been replaced by a script which installs the internal docs parser as a package and manually packages and pushes up the built docs to Artifactory.

The script was added to a new `tooling` folder, necessary steps were added to `Makefile` and release workflow was updated.

## Additional Context

Changes based on these PRs in `dara`:
- [PR 42](https://github.com/causalens/dara/pull/42)
- [PR 44](https://github.com/causalens/dara/pull/44)
- [PR 47](https://github.com/causalens/dara/pull/47)

## How Has This Been Tested?

Cannot be tested yet until new JFrog secrets are added to repo. Will test pushing a dev tag from this branch; unsure if this will suitably test the version bump commit when doing release from `main` branch as it has branch protection. 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (non-breaking change which improves functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactor (moving code around, general refactoring improvements)
- [ ] Documentation (documentation)

## PR Checklist:

- [x] I have reviewed my own PR? (review the PR as you are reviewing someone else's PR)
- [x] I have implemented all requirements? (see JIRA, project documentation)
- [x] I am affecting someone else's work? If so: I have added those people as reviewers?
- [x] I have added comments to all the bits that are hard to follow
- [x] At a bare minimum, I tested this manually
- [ ] I have added unit test(s)
- [ ] Is this a bugfix? If so have I added a regression test?
- [x] Does this PR satisfy the [one PR, one concern](https://medium.com/@fagnerbrack/one-pull-request-one-concern-e84a27dfe9f1) rule?
- [ ] If needed, documentation has been added/updated?
- [x] I have updated the appropriate changelog with a line for my changes
- [ ] If this PR breaks reproducibility, have I added a note in the changelog explaining the break in reproducibility
      and its extent?

### CODE REVIEW

File: .github/CODEOWNERS [hunk 1]
```diff
@@ -8,3 +8,4 @@
 /cai_causal_graph/ @causalens/causal-ai-team @causalens/mle
 /docs/ @causalens/causal-ai-team @causalens/mle
 /tests/ @causalens/causal-ai-team @causalens/mle
+/tooling/ @causalens/platform-team
```
Comment (andrew-causalens) [inline .github/CODEOWNERS:11]: @sam-causalens are you happy with this? Figured platform team should be aware of any script changes.

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -113,17 +111,19 @@ jobs:
...
-          artifactory-password: ${{ secrets.JF_PASSWORD }}
-          artifactory-username: ${{ secrets.JF_USERNAME }}
-          artifactory-url: https://causalens.jfrog.io/artifactory
-          package-name: cai-causal-graph
-          package-paths: cai_causal_graph
-          version: ${{ steps.extract_tag.outputs.new_version }}
+        env:
+          ARTIFACTORY_URL: https://causalens.jfrog.io/artifactory
+          ARTIFACTORY_USERNAME: ${{ secrets.JF_USERNAME }}
+          ARTIFACTORY_PASSWORD: ${{ secrets.JF_PASSWORD }}
```
Comment (andrew-causalens) [inline .github/workflows/release.yml:120]: @sam-causalens just need these secrets added to the repo.

Comment (andrew-causalens) [reply .github/workflows/release.yml:120]: We also need `GH_ACTIONS_KEY`

File: .github/CODEOWNERS (hunk 1 above)
Comment (sam-causalens) [reply .github/CODEOWNERS:11]: Yeah, this is fine.

File: tooling/scripts/docs-upload.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+from requests.auth import HTTPBasicAuth
+
+os.mkdir('__docs')
+
+package = os.environ['PACKAGE']
+version = os.environ['VERSION']
+
+# Build reference docs
+if os.environ.get('BUILD_REFERENCE') == 'true':
+    package_paths = os.environ.get('PACKAGE_PATHS', '').strip().split('\n')
```
Comment (andrew-causalens) [inline tooling/scripts/docs-upload.py:32]: @sam-causalens I could not test it without the secrets but I figured this was still OK even though we only have a single path (unlike the multiple paths for the `dara` monorepo). Just flagging as I was unsure here. But given we will have no new-line then the final split just won't happen.

File: .github/workflows/release.yml (hunk 2 above)
Comment (sam-causalens) [reply .github/workflows/release.yml:120]: I've added the jFrog ones for you, but I don't feel comfortable adding the GH_ACTIONS_KEY. I have check and can't seem to create one that is scoped to just this repo so I don't want to risk that in a public repository. I think you'll have to tweak your build process a bit to line up with the Dara Teams process where they push a PR with the version bump and changelog update rather than having it work automatically.

**Review by sam-causalens — COMMENTED** [review]
Changes look okay, but we'll need to remove the use of GH_ACTIONS_KEY as it's not safe in a public repo.

**Comment by max-causalens** [comment]
@andrew-causalens Sorry this may be a naive question but why don't we just redirect / link the user to the open source docs instead of duplicating it in the internal?

**Comment by andrew-causalens** [comment]
> @andrew-causalens Sorry this may be a naive question but why don't we just redirect / link the user to the open source docs instead of duplicating it in the internal?

The doc website is built from the packaged docs hosted on artifactory. I discussed this with @sam-causalens when we first started planning how to host the public docs and decided this was the simplest. This is also what Dara does too.

File: Makefile [hunk 4]
```diff
@@ -102,6 +103,13 @@ build-sdist:
 remove-generic-wheel:
 	rm -f dist/*-py3-*.whl
 
+# Package and publish the docs to causaLens' artifactory
+package-and-publish-docs:
+	poetry source add causalens https://causalens.jfrog.io/artifactory/api/pypi/python-open-source/simple/
+	poetry config http-basic.causalens $${ARTIFACTORY_USERNAME} $${ARTIFACTORY_PASSWORD}
```
Comment (sam-causalens) [inline Makefile:109]: @andrew-causalens I think the $$ is causing your issue with the build. Can you try changing it to a single one instead, e.g. the publish step below taking pypi token?

Comment (andrew-causalens) [reply Makefile:109]: Here is the same block in `dara` btw: https://github.com/causalens/dara/blob/master/Makefile#L79

Comment (sam-causalens) [reply Makefile:109]: Okay, I tried setting the creds again. Give it another try with the double $$ and see if it was a typo there.

Comment (andrew-causalens) [reply Makefile:109]: Thanks @sam-causalens! 🙏 . Kicked it off and will try single $ if it fails.

Comment (andrew-causalens) [reply Makefile:109]: Just to confirm it still [fails](https://github.com/causalens/cai-causal-graph/actions/runs/5925524641/job/16065113134) with $$:
```
poetry add --source=causalens docs-builder@~0.2.0
Source (causalens): Authorization error accessing https://causalens.jfrog.io/artifactory/api/pypi/python-open-source/simple/docs-builder/

Could not find a matching version of package docs-builder
```

Will try $ now.

Comment (andrew-causalens) [reply Makefile]: @sam-causalens it [fails](https://github.com/causalens/cai-causal-graph/actions/runs/5925564928/job/16065233038) with single $ too:
```
poetry add --source=causalens docs-builder@~0.2.0
Source (causalens): Authorization error accessing https://causalens.jfrog.io/artifactory/api/pypi/python-open-source/simple/docs-builder/

Could not find a matching version of package docs-builder
```

Can you check on JFrog to see if the requests are even making it to it and potentially get some further info from the logs?

Comment (sam-causalens) [reply Makefile]: I tried recreating the secrets in github altogether. Can you switch back to $$ and try again?

Comment (andrew-causalens) [reply Makefile:109]: Still [fails](https://github.com/causalens/cai-causal-graph/actions/runs/5926898619/job/16069224948) with new secrets and back to $$.

Comment (andrew-causalens) [reply Makefile:109]: All good now. Thanks!

File: .github/workflows/release.yml (hunk 2 above)
Comment (andrew-causalens) [reply .github/workflows/release.yml:120]: Discussed offline, actioned, and tested.

**Review by sam-causalens — APPROVED** [review]

**Comment by andrew-causalens** [comment]
> Looks good to me. You might want to revert the poetry.lock and pyproject.toml changes related to dev releases? But, no worries if not, the next release will override these.

I plan to release 0.1.1 from `main` so will leave.
